File: 23/A.hs
import Data.List
import Data.Maybe
import Debug.Trace
-- import qualified Data.Set as Set
import Data.Graph

isValidNextNeighbor d [x, y] [-1,  0] = d !! y !! (x-1) == '.' || d !! y !! (x-1) == '<'
isValidNextNeighbor d [x, y] [ 1,  0] = d !! y !! (x+1) == '.' || d !! y !! (x+1) == '>'
isValidNextNeighbor d [x, y] [ 0, -1] = d !! (y-1) !! x == '.' || d !! (y-1) !! x == '^'
isValidNextNeighbor d [x, y] [ 0,  1] = d !! (y+1) !! x == '.' || d !! (y+1) !! x == 'v'

neighbors d l =
    let [ width, height ] = [ length $ head d, length d ] 
        [ x, y ] = [ l `mod` width, l `div` width ]
    in [ (x+dx) * width + (y+dy) | 
        let [mx, my] = [ length $ head d, length d ], 
        dx <- [-1..1], dy <- [-1..1], 
        abs(dx) + abs(dy) == 1,
        x+dx >= 0, y+dy >= 0, x+dx < mx, y+dy < my,
        isValidNextNeighbor d [x, y] [dx, dy]
    ]

get d l = 
    let [ width, height ] = [ length $ head d, length d ] 
        [ x, y ] = [ l `mod` width, l `div` width ]
    in d !! y !! x

-- paths :: [[Char]] -> [Int] -> [[[Int]]]
-- paths d start = paths' [start] []
--     where
--         paths' [] visited = [visited]
--         paths' (p:xs) visited = 
--             let next = filter (\n -> not $ n `elem` visited) $ neighbors d p
--             in paths' next (p:visited) ++ paths' xs (p:visited)

-- bfs g ts = f ts b [] Set.empty
--   where
--     f x fw bw s
--       | Set.member x s = fw bw s
--       | otherwise      = x : fw (neighbors g x : bw) (Set.insert x s)

--     b [] _ = []
--     b qs s = foldl (foldr f) b qs [] s

makeVertices d p = mkv [p] []
    where 
        mkv [] _ = []
        mkv (p:xs) visited = 
            let nbrs = filter (\v -> not $ v `elem` visited) $ neighbors d p
            in map (\v -> (p, v)) (trace (show nbrs) nbrs) ++ mkv nbrs (p:visited)

main = do
    contents <- fmap lines $ readFile "test.txt"
    let [ width, height ] = [ length $ head contents, length contents ]
    let start = fromJust $ elemIndex '.' $ head contents
    let end = (fromJust $ elemIndex '.' $ last contents) + (height - 1) * width
    print start
    print $ makeVertices contents start
    print $ map (get contents) $ neighbors contents 24
    -- print $ sortOn length $ bfs contents start
    -- let path = last $ sortOn length $ paths contents start
    -- print path
    -- putStrLn $ unlines [[if [x, y] `elem` path then 'O' else contents !! y !! x | x <- [0..(length $ head contents) - 1]] | y <- [0..length contents]]

File: 23/A.js
const { readFileSync } = require("fs");

const contents = readFileSync("input.txt", { encoding: "utf-8" });

const grid = contents.split("\n");
const list = grid.join("");

const width = grid[0].length;
const height = grid.length;

const start = grid[0].indexOf(".") + 0 * width;
const end = grid[height-1].indexOf(".") + (height-1) * width;


// const neighbors = v => [v-1, v+1, v-width, v+width].filter(v => v >= 0 && v < width * height && list[v] != "#");

function *neighbors(v) {
	if (v-1 >= 0 				&& ".<".includes(list[v-1])) 		yield v-1;
	if (v+1 < width*height 		&& ".>".includes(list[v+1])) 		yield v+1;
	if (v-width >= 0 			&& ".^".includes(list[v-width])) 	yield v-width;
	if (v+width < width*height 	&& ".v".includes(list[v+width])) 	yield v+width;
}

function bfs(p, visited) {
	let n = [...neighbors(p)].filter(v => !visited.has(v));
	while (n.length === 1) {
		visited.add(p);
		p = n[0];
		n = [...neighbors(p)].filter(v => !visited.has(v));
	}
	visited.add(p);
	// console.log(c, n, p, visited);
	let maxNested = 0;
	for (const x of n) {
		// const newSet = new Set(visited);
		maxNested = Math.max(maxNested, bfs(x, new Set([p])));
	}
	return visited.size-1 + maxNested;
}

console.log(bfs(start, new Set()));